src/v/datalake/arrow_writer.cc

```diff
@@ -0,0 +1,241 @@
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::time_type& /*t*/) {
+        return std::make_unique<scalar_converter<arrow::Time64Type>>(
+          arrow::time64(arrow::TimeUnit::MICRO));
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::timestamp_type& /*t*/) {
+        return std::make_unique<scalar_converter<arrow::TimestampType>>(
+          arrow::timestamp(arrow::TimeUnit::MICRO));
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::timestamptz_type& /*t*/) {
+        // FIXME: time zone?
+        return std::make_unique<scalar_converter<arrow::TimestampType>>(
+          arrow::timestamp(arrow::TimeUnit::MICRO, "UTC"));
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::string_type& /*t*/) {
+        return std::make_unique<scalar_converter<arrow::StringType>>(
+          arrow::utf8());
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::uuid_type& /*t*/) {
+        // FIXME: there's an arrow extension for a uuid logical type.
+        // Under-the-hood it's stored as a fixed(16)
+        return std::make_unique<scalar_converter<arrow::FixedSizeBinaryType>>(
+          arrow::fixed_size_binary(16));
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::fixed_type& t) {
+        return std::make_unique<scalar_converter<arrow::FixedSizeBinaryType>>(
+          arrow::fixed_size_binary(static_cast<uint32_t>(t.length)));
+    }
+    std::unique_ptr<converter_interface>
+    operator()(const iceberg::binary_type& /*t*/) {
+        return std::make_unique<scalar_converter<arrow::BinaryType>>(
+          arrow::binary());
+    }
+};
+
+struct_converter::struct_converter(const iceberg::struct_type& schema) {
+    _fields.reserve(schema.fields.size());
+    for (const auto& field : schema.fields) {
+        auto child = std::visit(converter_builder_visitor{}, field->type);
+        _fields.push_back(child->field(field->name, field->required));
+        _child_converters.push_back(std::move(child));
+    }
+    _arrow_data_type = arrow::struct_(_fields);
+}
+
+list_converter::list_converter(const iceberg::list_type& schema) {
+    auto child = std::visit(
+      converter_builder_visitor{}, schema.element_field->type);
+    _child_field = child->field(
+      schema.element_field->name, schema.element_field->required);
+    _child_converter = std::move(child);
+}
+
+map_converter::map_converter(const iceberg::map_type& schema) {
+    auto key = std::visit(converter_builder_visitor{}, schema.key_field->type);
+    auto value = std::visit(
+      converter_builder_visitor{}, schema.value_field->type);
+
+    _key_field = key->field(schema.key_field->name, schema.key_field->required);
+    _value_field = value->field(
+      schema.value_field->name, schema.value_field->required);
+
+    _key_converter = std::move(key);
+    _value_converter = std::move(value);
```

Comment:
Curious what the motivation is for keeping these around? Is it to ensure lifetime of the underlying fields?

Given everything is shared, it kind of seems like the visitors could have returned fields directly, rather than wrapping in converters. Or will the converters play double duty with the data path too and require some longer lived state (builders?)

Reply:
You're right, in this PR they're not needed after this point, but in the next PR they will be used to convert the actual values.